Framework/Algorithms/inc/MantidAlgorithms/SpectrumAlgorithm.h

```diff
@@ -0,0 +1,170 @@
+      operation(std::get<S>(getters)(workspace, indexSet[i])...);
+      progress.report(name());
+      PARALLEL_END_INTERUPT_REGION
+    }
+    PARALLEL_CHECK_INTERUPT_REGION
+
+    ifEventWorkspaceClearMRU(workspace);
+  }
+
+  /// Templated function used as 'compile-time conditional', together with
+  /// specialization. See there for details.
+  template <class WS> void ifEventWorkspaceClearMRU(const WS &workspace) {
+    UNUSED_ARG(workspace);
+  }
+
+protected:
+  /// Dummy struct holding compile-time flags to for_each().
+  // A strongly typed enum used in non-type variadic template arguments would
+  // have been the other candidate. However, combining flags from multiple enums
+  // would not have been possible, so all flags would have been required to be
+  // in the same scoped enum, and we want to avoid noisy flag definitions like
+  // Flags::IndicesFromProperty or Flags::SkipMasked.
+  struct Indices {
+    /// Flag: Include only indices specified via properties in for_each.
+    struct FromProperty {};
+  };
+
+  /** Provides a mechanism for looping over spectra in a workspace.
+   *
+   * This variant works with a single workspace and can to in-place modification
+   * of spectra or event lists in the workspace. Threading and progress
+   * reporting is handled internally.
+   * @tparam Flags Variable number of flags, see struct Indices.
+   * @param workspace Workspace to work with.
+   * @param getters Tuple of getters, pointers to methods of workspace, defaults
+   *     defined for convenience in namespace Getters.
+   * @param operation Callable that is executed for all spectra (etc.). Results
+   *     from getters are passed as arguments. */
+  template <class... Flags, class WS, class... Args, class OP>
+  void for_each(WS &workspace, std::tuple<Args...> getters,
+                const OP &operation) {
+    // This comment explains some of the rationale and mechanism for the way
+    // templates are used in this and other variants of for_each().
+    // For several variants of for_each() we require a variable number of
+    // arguments for more than one entity, e.g., getters for the input workspace
+    // and getters for the output workspace. For both of them we would thus like
+    // to use a variadic template. However, we cannot make all those getters
+    // direct arguments to for_each(), because the compiler then cannot tell
+    // were the boundary between the two parameter packs is. Therefore getters
+    // are packed into tuples.
+    // At some point we need to extract the getters from the tuple. To this end,
+    // we use the gens template, which recursively constructs a sequence of
+    // indices which are gen used with std::get<>.
+    // The Flags template parameter is used for passing in flags known at
+    // compile time.
+    for_each<Flags...>(workspace, getters,
+                       typename gens<sizeof...(Args)>::type(), operation);
+  }
+
+  void declareSpectrumIndexSetProperties();
```

Comment:
Can we change the name here to `declareWorkspaceIndexSetProperties` to avoid mixing the already ambiguous terms?